File: jsrc/machine/SxxMachine/ListViewTerm.java
package SxxMachine;

import java.util.ArrayList;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.List;

/**
 * List term that allows tail addition to the list without recreating the list.
 * This class behaves like a view for underlying List instance.
 */
public class ListViewTerm extends ListTerm {
	// the class is necessary to make ListTerm.isImmutable() to return false
	private final static Term NOT_IMMUTABLE = new Term() {
		@Override
		public boolean unify(Term t, Trail trail) {
			return false;
		}

		@Override
		public boolean isImmutable() {
			return false;
		}

		@Override
		public String name() {
			return "";
		}

		@Override
		public void toString(StringBuilder sb) {

		}

		@Override
		public void toQuotedString(StringBuilder sb) {

		}

		@Override
		public int compareTo(Term o) {
			return 0;
		}
	};
	/** the list of terms */
	private final List<Term> list;
	/** index of element in the {@link SxxMachine.ListViewTerm#list} represented by this instance*/
	private final int index;
	/** holds next {@link SxxMachine.ListViewTerm} instance,
	 * the value is lazy initialized */
	private Term next = null;

	public ListViewTerm(Term head) {
		super(head, NOT_IMMUTABLE);
		this.list = new ArrayList<>();
		list.add(head);
		this.index = 0;
	}

	private ListViewTerm(List<Term> list, int index) {
		super(list.get(index), NOT_IMMUTABLE); // makes isImmutable to return false
		this.list = list;
		this.index = index;
	}

	@Override
	public ListTerm add(Term term){
		list.add(term);
		return this;
	}

	@Override
	public Term cdr() {
		if (next!=null) {
			return next;
		} else if (index + 1 < list.size()) {
			return next = new ListViewTerm(list, index + 1);
		} else {
			return Prolog.Nil;
		}
	}

	@Override
	protected Term copy(IdentityHashMap<VariableTerm, VariableTerm> copyHash) {

		Term result = Prolog.Nil;
		for(int i=list.size()-1; i>=index; i--){
			result = new ListTerm(list.get(i).copy(copyHash), result);
		}
		return result;
	}

	@Override
	public Term arg(int nth) {
		return list.get(index+nth);
	}

	@Override
	public int length() {
		return list.size() - index;
	}

	@Override
	public List toJava() {
		return Collections.unmodifiableList(list.subList(index, list.size()));
	}

	@Override
	public boolean isGround() {
		for(int i=list.size()-1; i>=index; i--){
			if (!list.get(i).isGround()){
				return false;
			}
		}
		return true;
	}
}

File: src/machine/SxxMachine/ListViewTerm.cpp
using namespace std;

#include "ListViewTerm.h"
#include "Trail.h"
#include "Term.h"
#include "OpVisitor.h"
#include "Prolog.h"
#include "TermData.h"

namespace SxxMachine
{

	int ListViewTerm::MutableMarkerForLists::type()
	{
		return TYPE_LIST;
	}

	bool ListViewTerm::MutableMarkerForLists::unifyImpl(Term *t, Trail *trail)
	{
		return false;
	}

	bool ListViewTerm::MutableMarkerForLists::isImmutable()
	{
		return false;
	}

	wstring ListViewTerm::MutableMarkerForLists::name()
	{
		return "";
	}

	void ListViewTerm::MutableMarkerForLists::toStringImpl(int printFlags, StringBuilder *sb)
	{
	}

	int ListViewTerm::MutableMarkerForLists::compareTo(Term *o)
	{
		return 0;
	}

	int ListViewTerm::MutableMarkerForLists::termHashCodeImpl()
	{
		return System::identityHashCode(this);
	}

	bool ListViewTerm::MutableMarkerForLists::equalsTerm(Term *obj, OpVisitor *comparator)
	{
		return this == obj;
	}

	wstring ListViewTerm::pprint()
	{
		return ListTerm::toQuotedString();
	}

	bool ListViewTerm::isImmutable()
	{
		return false;
	}

Term *const ListViewTerm::NOT_IMMUTABLE = new MutableMarkerForLists();

	ListViewTerm::ListViewTerm(Term *head) : ListTerm(head, NOT_IMMUTABLE), list(vector<>()), index(0)
	{
		//this.argz = VA(head, NOT_IMMUTABLE);
		this->list.push_back(head);
	}

	ListViewTerm::ListViewTerm(vector<Term*> &list, int index) : ListTerm(list.get(index), NOT_IMMUTABLE), list(list), index(index)
	{
		//this.argz = VA(list.get(index), NOT_IMMUTABLE);//super(list.get(index), NOT_IMMUTABLE); // makes isImmutable to return false
	}

	ListTerm *ListViewTerm::add(Term *term)
	{
		this->list.push_back(term);
		return this;
	}

	Term *ListViewTerm::cdr()
	{
		if (this->next != nullptr)
		{
			return this->next;
		}
		else if (this->index + 1 < this->list.size())
		{
			return this->next = new ListViewTerm(this->list, this->index + 1);
		}
		else
		{
			return Prolog::Nil;
		}
	}

	Term *ListViewTerm::copyImpl(IdentityHashMap<any, Term*> *copyHash, int deeply)
	{
		Term *result = Prolog::Nil;
		for (int i = this->list.size() - 1; i >= this->index; i--)
		{
			result = TermData::CONS(this->list[i]->copy(copyHash, deeply), result);
		}
		return result;
	}

	Term *ListViewTerm::nth0(int nth)
	{
		return this->list[this->index + nth];
	}

	int ListViewTerm::length()
	{
		return this->list.size() - this->index;
	}

	vector ListViewTerm::toJava()
	{
		return Collections::unmodifiableList(this->list.subList(this->index, this->list.size()));
	}

	bool ListViewTerm::isGround()
	{
		for (int i = this->list.size() - 1; i >= this->index; i--)
		{
			if (!this->list[i]->isGround())
			{
				return false;
			}
		}
		return true;
	}

	Term *ListViewTerm::car()
	{
		// TODO Auto-generated method stub
		return nth0(0);
	}

	void ListViewTerm::setCar(Term *t)
	{
		Prolog::Break("SETCAR");
		argz[0] = t;

	}

	void ListViewTerm::setCdr(Term *t)
	{
		Prolog::Break("SETCDR");
		argz[1] = t;
	}

	std::vector<Term*> ListViewTerm::args()
	{
		// TODO Auto-generated method stub
		Prolog::Break("ARGS");
		return argz;
	}

	wstring ListViewTerm::name()
	{
		return Prolog::SYM_DOT->name();
	}

	Term *ListViewTerm::toClone() throw(CloneNotSupportedException)
	{
		return new ListViewTerm(list,index);
	}

	int ListViewTerm::arityOrType()
	{
		return type();
	}
}
